### PR TITLE
[Profiler] Automatic pprof profile uploads to Google Cloud Profiler.

### PR DESCRIPTION
**Pull request by SaveTheRbtz**
This diffs adds an ability to upload AutoProfiler pprof files to [Cloud Profiler](https://cloud.google.com/profiler/docs/about-profiler).

For this to work VM should be running in GCE, have access to `cloudprofiler.googleapis.com:443`, and have the `https://www.googleapis.com/auth/monitoring.write` scope.  Also, one should check their "Create offline requests per day" quota.

Note that, this functionality is currently hidden behind both `-profiler-enabled` and `-profile-uploader-enabled` feature flags.

While here, also:
* Removed Mutex profiling, since [Block profiling is better](https://github.com/DataDog/go-profiler-notes/blob/main/block.md) for our usecase. 
* Switched from heap to alloc profile.  These only differ in the default sample name but identical otherwise.

Issue: https://github.com/dapperlabs/flow-go/issues/6310

Followup diffs:
- [ ] Add metrics.
- [ ] Add localnet support for uploading pprof data.

### CODE REVIEW

File: cmd/scaffold.go [hunk 1]
```diff
@@ -19,6 +19,9 @@ import (
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/rs/zerolog"
 	"github.com/spf13/pflag"
+	"google.golang.org/api/option"
+
```
Comment (pattyshack) [inline cmd/scaffold.go:23]: rm newline

File: cmd/scaffold.go [hunk 2]
```diff
@@ -530,12 +539,58 @@ func (fnb *FlowNodeBuilder) initMetrics() {
 	}
 }
 
+func (fnb *FlowNodeBuilder) createUploader(client *gcemd.Client, opts ...option.ClientOption) (profiler.Uploader, error) {
```
Comment (pattyshack) [inline cmd/scaffold.go:542]: rename to createGCEProfileUploader

File: cmd/scaffold.go [hunk 3]
```diff
@@ -530,12 +539,58 @@ func (fnb *FlowNodeBuilder) initMetrics() {
...
+	fnb.Logger.Info().Msgf("creating pprof profile uploader with params: %+v", params)
+
+	return profiler.NewUploader(fnb.Logger, params, opts...)
+}
+
 func (fnb *FlowNodeBuilder) initProfiler() {
 	// note: by default the Golang heap profiling rate is on and can be set even if the profiler is NOT enabled
 	runtime.MemProfileRate = fnb.BaseConfig.profilerMemProfileRate
 
+	var err error
```
Comment (pattyshack) [inline cmd/scaffold.go:578]: move this out of initProfiler into a separate function, say createProfileUploader (also see my above renaming comment)

File: cmd/scaffold.go [hunk 4]
```diff
@@ -530,12 +539,58 @@ func (fnb *FlowNodeBuilder) initMetrics() {
...
+		Commit:    build.Commit(),
+		Instance:  instance,
+	}
+
+	fnb.Logger.Info().Msgf("creating pprof profile uploader with params: %+v", params)
+
+	return profiler.NewUploader(fnb.Logger, params, opts...)
+}
+
 func (fnb *FlowNodeBuilder) initProfiler() {
```
Comment (pattyshack) [inline cmd/scaffold.go:574]: not specific to your PR, but ignore errors (only logging warning) seems kind of terrible.  we should propagate the error whenever possible ...

File: module/profiler/profiler.go [hunk 5]
```diff
@@ -135,18 +157,6 @@ func (p *AutoProfiler) pprofBlock(w io.Writer) error {
 	}
 }
 
-func (p *AutoProfiler) pprofMutex(w io.Writer) error {
```
Comment (pattyshack) [inline module/profiler/profiler.go:138]: why was this removed?  if the gce uploader don't support this, maybe just don't upload it

File: module/profiler/profiler.go [hunk 6]
```diff
@@ -116,11 +138,11 @@ func newProfileFunc(name string) profileFunc {
 	}
 }
 
-func (p *AutoProfiler) pprofHeap(w io.Writer) error {
+func (p *AutoProfiler) pprofAllocs(w io.Writer) error {
```
Comment (pattyshack) [inline module/profiler/profiler.go:141]: if allocs and heap profile different things, we should profile both.

File: module/profiler/uploader.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+	log = log.With().Str("component", "profile_uploader").Logger()
+
+	defaultOpts := []option.ClientOption{
+		option.WithEndpoint(apiAddress),
+		option.WithScopes(scope),
+		option.WithUserAgent("OfflinePprofUploader"),
+		option.WithTelemetryDisabled(),
+		option.WithGRPCDialOption(
+			grpc.WithDefaultCallOptions(
+				grpc.MaxCallSendMsgSize(maxMsgSize),
```
Comment (pattyshack) [inline module/profiler/uploader.go:59]: you realize this potentially means a 1GB protobuf message, right?

File: module/profiler/profiler.go (hunk 5 above)
Comment (SaveTheRbtz) [reply module/profiler/profiler.go:138]: Will mention this in the description: "mutex" is almost a subset of "block" but less useful (it tracks where locks are held instead of where code is waiting (blocked) on acquiring the lock).

File: module/profiler/profiler.go (hunk 6 above)
Comment (SaveTheRbtz) [reply module/profiler/profiler.go:141]: will mention this in the description: alloc and heap are the same. the only difference is which sample is marked as "default".

File: module/profiler/uploader.go (hunk 7 above)
Comment (SaveTheRbtz) [reply module/profiler/uploader.go:59]: yep. Our traces right now are about ~60MiB, so i've given it ~ an order of magnitude of potential growth.  That said, it is still an improvement to the default 2GiB: https://github.com/grpc/grpc-go/blob/13b378bc458535e6da1693fc67e09a4951cd3259/clientconn.go#L97

Comment (pattyshack) [reply module/profiler/uploader.go:59]: o.  interesting.  this default used to be 4MB a few years back (I had to bump it up to 6mb by default at dbx cuz of block size).  i guess everyone got tired of manually increasing it

**Review by pattyshack — APPROVED** [review]

**Review by AndrewM-SDET — APPROVED** [review]
Approving. There is one error message typo but code itself looks good.

**Comment by SaveTheRbtz** [comment]
bors merge
